### PR TITLE
Versal Platform Banner, RPMB key control, HUK driver and FPGA programming.

### PR DESCRIPTION
**Pull request by ldts**

### CODE REVIEW

**Comment by ldts** [comment]
@nathan-menhorn jfyi

**Comment by ldts** [comment]
after this, functionality should be on par with the [Foundries.io tree](https://github.com/foundriesio/optee_os/tree/3.18+fio)  (maybe some minor stuff will pop-up but I dont expect anything major)

**Comment by nathan-menhorn** [comment]
Thanks @ldts

**Comment by nathan-menhorn** [comment]
@jenswi-linaro @jbech-linaro any updates on this PR? Thanks. It seems it would be good to get these in before the 3.20 release.

File: core/drivers/versal_huk.c [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+ * will BE swap the values.
+ *
+ * This way the test key below corresponds to the byte array 0xf8, 0x78, 0xb8,
+ * 0x38, 0xd8, 0x58, 0x98, 0x18, 0xe8, 0x68, ....
+ *
+ * NOTICE: This hardcoded value in DEVEL_KEY could have just been zeroes as done
+ * in the weak implementation found in otp_stubs.c.
+ */
+#define DEVEL_KEY { \
+0xf878b838, 0xd8589818, 0xe868a828, 0xc8488808, \
```
Comment (jenswi-linaro) [inline core/drivers/versal_huk.c:97]: Please indent with a tab or two. Same for the other unindented #define values below.

File: core/drivers/versal_huk.c [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+	free(init);
+	free(p.buf);
+	memset(&cmd, 0, sizeof(cmd));
+	if (ret)
+		return ret;
+
+	versal_mbox_alloc(aad_len, aad_data, &p);
+	cmd.data[0] = API_ID(VERSAL_AES_UPDATE_AAD);
+	reg_pair_from_64(virt_to_phys(p.buf), &cmd.data[2], &cmd.data[1]);
+	cmd.data[3] = p.len % 16 ? p.alloc_len : p.len;
```
Comment (jenswi-linaro) [inline core/drivers/versal_huk.c:236]: Please use a regular if statement instead of the ternary operator.

File: core/arch/arm/plat-versal/main.c [hunk 3]
```diff
@@ -56,3 +66,35 @@ void console_init(void)
...
+
+	if (io_read32(plm_rtca + VERSAL_AHWROT_REG) == VERSAL_AHWROT_SECURED)
+		ahwrot = true;
+
+	if (io_read32(plm_rtca + VERSAL_SHWROT_REG) == VERSAL_SHWROT_SECURED)
+		shwrot = true;
+
+	IMSG("Hardware Root of Trust:\t"
+	     "Asymmetric %s Enabled, Symmetric %s Enabled",
+	     ahwrot ? "" : "NOT", shwrot ? "" : "NOT");
```
Comment (jenswi-linaro) [inline core/arch/arm/plat-versal/main.c:96]: This is more complicated than necessary, you could assign the correct strings above instead.

File: core/arch/arm/plat-versal/conf.mk [hunk 4]
```diff
@@ -70,3 +70,12 @@ CFG_VERSAL_PM ?= y
...
 CFG_VERSAL_PUF ?= y
+
+# Enable Hardware Unique Key driver
+CFG_VERSAL_HUK ?= y
+# AES-GCM supported key sources for HUK:
+#     6  : eFUSE USR 0
+#     7  : eFuse USR 1
+#    11  : PUF KEK
+#    12  : AES User Key 0 (devel)
+CFG_VERSAL_HUK_KEY ?= 12
```
Comment (etienne-lms) [inline core/arch/arm/plat-versal/conf.mk:81]: check against valid values? (preferrable here and at least in the embedded code)

Comment (jforissier) [reply core/arch/arm/plat-versal/conf.mk:81]: +1, I suggest:
```make
CFG_VERSAL_HUK_KEY ?= 12
ifneq ($(CFG_VERSAL_HUK_KEY),$(filter 6 7 11 12,$(firstword $(CFG_VERSAL_HUK_KEY))))
$(error Invalid value: CFG_VERSAL_HUK_KEY=$(CFG_VERSAL_HUK_KEY))
endif
```

Comment (ldts) [reply core/arch/arm/plat-versal/conf.mk:81]: thanks a lot @jforissier for the detail!

File: core/drivers/versal_huk.c [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+
+	switch (src) {
+	case VERSAL_AES_EFUSE_USER_KEY_0:
+		if (versal_efuse_read_sec_ctrl(&ctrl))
+			panic();
+
+		if (ctrl.user_key0_wr_lk)
+			*secure = true;
+		else
+			*secure = false;
```
Comment (jforissier) [inline core/drivers/versal_huk.c:131]: I think the compiler will do the right thing with just `*secure = ctrl.user_key0_wr_lk;`.

File: core/drivers/versal_huk.c [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+		return true;
+
+	case VERSAL_AES_EFUSE_USER_KEY_1:
+		if (versal_efuse_read_sec_ctrl(&ctrl))
+			panic();
+
+		if (ctrl.user_key1_wr_lk)
+			*secure = true;
+		else
+			*secure = false;
```
Comment (jforissier) [inline core/drivers/versal_huk.c:141]: Same here

File: core/drivers/versal_huk.c [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+		if (versal_efuse_read_puf_sec_ctrl(&puf_ctrl))
+			panic();
+
+		if (versal_puf_regenerate(&puf_data, &cfg))
+			panic();
+
+		if (puf_ctrl.puf_syn_lk)
+			*secure = true;
+		else
+			*secure = false;
```
Comment (jforissier) [inline core/drivers/versal_huk.c:154]: Same here

File: core/drivers/versal_huk.c [hunk 8]
```diff
@@ -0,0 +1,339 @@
...
+	memset(&cmd, 0, sizeof(cmd));
+
+	return ret;
+}
+
+TEE_Result tee_otp_get_hw_unique_key(struct tee_hw_unique_key *hwkey)
+{
+	uint32_t dna[EFUSE_DNA_LEN / sizeof(uint32_t)] = { 0 };
+	uint8_t enc_data[64] = { 0 };
+	uint8_t sha[48] = { 0 };
```
Comment (jforissier) [inline core/drivers/versal_huk.c:304]: `= { };`

**Review by jforissier — COMMENTED** [review]
Minor comments on "drivers: versal: hardware unique key", in any case:

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Review by jforissier — COMMENTED** [review]
For commit "plat-versal: print platform information":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Review by jforissier — COMMENTED** [review]
For "plat-versal: allow RPMB init only on secured boards":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Review by jforissier — COMMENTED** [review]
For "plat-versal: program the FPGA during OP-TEE initialization":

`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Review by jforissier — COMMENTED** [review]
For "ci: versal: program FPGA":

`Reviewed-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/arch/arm/plat-versal/main.c [hunk 9]
```diff
@@ -56,3 +66,34 @@ void console_init(void)
...
+	uint8_t version = 0;
+
+	assert(plm_rtca);
+
+	if (versal_soc_version(&version)) {
+		EMSG("Failure to retrieve SoC version");
+		return TEE_ERROR_GENERIC;
+	}
+
+	IMSG("Platform Versal:\tSilicon Revision v%d", version);
```
Comment (etienne-lms) [inline core/arch/arm/plat-versal/main.c:86]: `%"PRIu8, version)`

File: core/arch/arm/plat-versal/platform_config.h [hunk 10]
```diff
@@ -14,6 +14,9 @@
 
 #if defined(PLATFORM_FLAVOR_generic)
 
+#define PLM_RTCA             0xF2014000
```
Comment (etienne-lms) [inline core/arch/arm/plat-versal/platform_config.h:17]: tabs

File: core/drivers/versal_huk.c [hunk 11]
```diff
@@ -0,0 +1,339 @@
...
+
+	versal_mbox_alloc(16, NULL, &p);
+	cmd.data[0] = API_ID(VERSAL_AES_ENCRYPT_FINAL);
+	reg_pair_from_64(virt_to_phys(p.buf), &cmd.data[2], &cmd.data[1]);
+	if (versal_mbox_notify(&cmd, NULL, NULL)) {
+		EMSG("AES_ENCRYPT_FINAL error");
+		ret = TEE_ERROR_GENERIC;
+	}
+	free(p.buf);
+	memset(&cmd, 0, sizeof(cmd));
```
Comment (etienne-lms) [inline core/drivers/versal_huk.c:295]: `memset_explicit()` here?

**Review by etienne-lms — COMMENTED** [review]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>` for commits:
"drivers: versal: hardware unique key" (preferrably with comments addressed)
"plat-versal: allow RPMB init only on secured boards"
"plat-versal: program the FPGA during OP-TEE initialization"
"ci: versal: program FPGA"

Issues to fix in commit "plat-versal: print platform information".

File: core/drivers/versal_huk.c (hunk 8 above)
Comment (jforissier) [reply core/drivers/versal_huk.c:304]: `= { };` is quite clear too and having only one option is simpler...

Comment (etienne-lms) [reply core/drivers/versal_huk.c:304]: ok, let's sitck to the conding standards. I'll fix the stm32 occurences after next tag, for consistency.

**Comment by ldts** [comment]
will this make it to 3.20?

**Comment by jforissier** [comment]
> will this make it to 3.20?

Yes, the changes are local enough to Versal to be safe to merge now I think. Thanks!

**Pull request merged**